File: Python3.x/121-Best Time to Buy and Sell Stock.py
# dp & 内存优化
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if len(prices) < 2: return 0
        min_price, max_profit = prices[0], 0  # 当前最低价 和 当前最大收益
        for i in range(1, len(prices)):
            min_price = min(min_price, prices[i])  # 买入最低价
            max_profit = max(max_profit, prices[i] - min_price)  # 最大收益比较
        return max_profit

File: C++/121-Best_Time_to_Buy_and_Sell_Stock.cpp
// Brute Force
// Runtime: 1664 ms, faster than 5.00% of C++ online submissions for Best Time to Buy and Sell Stock.
// Memory Usage: 9.5 MB, less than 84.40% of C++ online submissions for Best Time to Buy and Sell Stock.
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int res = 0;
        for (int i = 0; i < prices.size(); ++i)
            for (int j = i + 1; j < prices.size(); ++j)
                res = max(res, prices[j] - prices[i]);
        return res;
    }
};


// Runtime: 4 ms, faster than 98.29% of C++ online submissions for Best Time to Buy and Sell Stock.
// Memory Usage: 9.6 MB, less than 74.31% of C++ online submissions for Best Time to Buy and Sell Stock.
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int max_profit = 0, min_price = INT_MAX;
        for (int i = 0; i < prices.size(); ++i) {
            min_price = min(min_price, prices[i]);
            max_profit = max(max_profit, prices[i] - min_price);
        }
        return max_profit;
    }
};
